module/VuFind/src/VuFind/Db/Service/OaiResumptionService.php

```
@@ -0,0 +1,123 @@
+use VuFind\Db\Table\DbTableAwareInterface;
+use VuFind\Db\Table\DbTableAwareTrait;
+use VuFind\Log\LoggerAwareTrait;
+
+/**
+ * Database service for OaiResumption.
+ *
+ * @category VuFind
+ * @package  Database
+ * @author   Sudharma Kellampalli <skellamp@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:database_gateways Wiki
+ */
+class OaiResumptionService extends AbstractDbService implements
+    DbTableAwareInterface,
+    LoggerAwareInterface,
+    OaiResumptionServiceInterface
+{
+    use DbTableAwareTrait;
+    use LoggerAwareTrait;
+
+    /**
+     * Remove all expired tokens from the database.
+     *
+     * @return void
+     */
+    public function removeExpired(): void
+    {
+        $this->getDbTable('oairesumption')->removeExpired();
+    }
+
+    /**
+     * Retrieve a row from the database based on primary key; return null if it
+     * is not found.
+     *
+     * @param string $token The resumption token to retrieve.
+     *
+     * @return ?OaiResumption
+     */
+    public function findToken($token): ?OaiResumptionEntityInterface
+    {
+        return $this->getDbTable('oairesumption')->findToken($token);
+    }
+
+    /**
+     * Create a new resumption token
+     *
+     * @param array $params Parameters associated with the token.
+     * @param int   $expire Expiration time for token (Unix timestamp).
+     *
+     * @return int          ID of new token
+     */
+    public function saveToken($params, $expire): int
+    {
+        $row = $this->createEntity()
+            ->setResumptionParameters($this->encodeParams($params))
+            ->setExpiry(\DateTime::createFromFormat('U', $expire));
+        try {
+            $this->persistEntity($row);
+        } catch (\Exception $e) {
+            $this->logError('Could not save token: ' . $e->getMessage());
+            return false;
```

The return type is `int`, so should it be changed to `?int`? And I guess in that case the PHPDoc shoud include an explanation.

---

I changed it to rethrow the exception instead of returning false.